File: app/src/main/kotlin/tech/mmarca/openvitals/domain/query/VitalsPeriodData.kt
package tech.mmarca.openvitals.domain.query

import tech.mmarca.openvitals.domain.model.BloodGlucoseEntry
import tech.mmarca.openvitals.domain.model.BloodPressureEntry
import tech.mmarca.openvitals.domain.model.BodyTempEntry
import tech.mmarca.openvitals.domain.model.DailyBloodPressurePoint
import tech.mmarca.openvitals.domain.model.DailyVitalPoint
import tech.mmarca.openvitals.domain.model.RespiratoryRateEntry
import tech.mmarca.openvitals.domain.model.SkinTemperatureEntry
import tech.mmarca.openvitals.domain.model.SpO2Entry
import tech.mmarca.openvitals.domain.model.Vo2MaxEntry
import tech.mmarca.openvitals.data.repository.VitalsPeriodMetric

data class VitalsPeriodData(
    val missingVitalsPermissions: Set<String> = emptySet(),
    /**
     * Overview metrics whose daily read blew its per-metric budget on this
     * range. Their lists arrive empty; the UI says so instead of blanking the
     * whole screen.
     */
    val timedOutMetrics: Set<VitalsPeriodMetric> = emptySet(),
    val bloodPressure: List<BloodPressureEntry> = emptyList(),
    val previousBloodPressure: List<BloodPressureEntry> = emptyList(),
    val baselineBloodPressure: List<BloodPressureEntry> = emptyList(),
    val spO2: List<SpO2Entry> = emptyList(),
    val previousSpO2: List<SpO2Entry> = emptyList(),
    val baselineSpO2: List<SpO2Entry> = emptyList(),
    val respiratoryRate: List<RespiratoryRateEntry> = emptyList(),
    val previousRespiratoryRate: List<RespiratoryRateEntry> = emptyList(),
    val baselineRespiratoryRate: List<RespiratoryRateEntry> = emptyList(),
    val bodyTemperature: List<BodyTempEntry> = emptyList(),
    val previousBodyTemperature: List<BodyTempEntry> = emptyList(),
    val baselineBodyTemperature: List<BodyTempEntry> = emptyList(),
    val vo2Max: List<Vo2MaxEntry> = emptyList(),
    val previousVo2Max: List<Vo2MaxEntry> = emptyList(),
    val baselineVo2Max: List<Vo2MaxEntry> = emptyList(),
    val bloodGlucose: List<BloodGlucoseEntry> = emptyList(),
    val previousBloodGlucose: List<BloodGlucoseEntry> = emptyList(),
    val baselineBloodGlucose: List<BloodGlucoseEntry> = emptyList(),
    val skinTemperature: List<SkinTemperatureEntry> = emptyList(),
    val previousSkinTemperature: List<SkinTemperatureEntry> = emptyList(),
    val baselineSkinTemperature: List<SkinTemperatureEntry> = emptyList(),
    // Non-day overview loads carry one aggregated point per local day instead of
    // raw entries, plus the window's true latest reading per metric for the cards.
    val bloodPressureDaily: List<DailyBloodPressurePoint> = emptyList(),
    val spO2Daily: List<DailyVitalPoint> = emptyList(),
    val respiratoryRateDaily: List<DailyVitalPoint> = emptyList(),
    val bodyTemperatureDaily: List<DailyVitalPoint> = emptyList(),
    val vo2MaxDaily: List<DailyVitalPoint> = emptyList(),
    val bloodGlucoseDaily: List<DailyVitalPoint> = emptyList(),
    val skinTemperatureDaily: List<DailyVitalPoint> = emptyList(),
    val latestBloodPressure: BloodPressureEntry? = null,
    val latestSpO2: SpO2Entry? = null,
    val latestRespiratoryRate: RespiratoryRateEntry? = null,
    val latestBodyTemperature: BodyTempEntry? = null,
    val latestVo2Max: Vo2MaxEntry? = null,
    val latestBloodGlucose: BloodGlucoseEntry? = null,
    val latestSkinTemperature: SkinTemperatureEntry? = null,
)

File: lib/domain/query/vitals_period_data.dart
import 'package:freezed_annotation/freezed_annotation.dart';

import '../../data/repository/contract/vitals_repository.dart';
import '../model/vitals_models.dart';

part 'vitals_period_data.freezed.dart';

@freezed
abstract class VitalsPeriodData with _$VitalsPeriodData {
  const factory VitalsPeriodData({
    @Default(<String>{}) Set<String> missingVitalsPermissions,
    // Non-day metrics whose daily read exceeded its budget (too large to read
    // raw over this range) — the card shows an "unavailable for this range"
    // state instead of "no readings". See VitalsRepositoryImpl.
    @Default(<VitalsPeriodMetric>{}) Set<VitalsPeriodMetric> timedOutMetrics,
    @Default(<BloodPressureEntry>[]) List<BloodPressureEntry> bloodPressure,
    @Default(<BloodPressureEntry>[])
    List<BloodPressureEntry> previousBloodPressure,
    @Default(<BloodPressureEntry>[])
    List<BloodPressureEntry> baselineBloodPressure,
    @Default(<SpO2Entry>[]) List<SpO2Entry> spO2,
    @Default(<SpO2Entry>[]) List<SpO2Entry> previousSpO2,
    @Default(<SpO2Entry>[]) List<SpO2Entry> baselineSpO2,
    @Default(<RespiratoryRateEntry>[])
    List<RespiratoryRateEntry> respiratoryRate,
    @Default(<RespiratoryRateEntry>[])
    List<RespiratoryRateEntry> previousRespiratoryRate,
    @Default(<RespiratoryRateEntry>[])
    List<RespiratoryRateEntry> baselineRespiratoryRate,
    @Default(<BodyTempEntry>[]) List<BodyTempEntry> bodyTemperature,
    @Default(<BodyTempEntry>[]) List<BodyTempEntry> previousBodyTemperature,
    @Default(<BodyTempEntry>[]) List<BodyTempEntry> baselineBodyTemperature,
    @Default(<Vo2MaxEntry>[]) List<Vo2MaxEntry> vo2Max,
    @Default(<Vo2MaxEntry>[]) List<Vo2MaxEntry> previousVo2Max,
    @Default(<Vo2MaxEntry>[]) List<Vo2MaxEntry> baselineVo2Max,
    @Default(<BloodGlucoseEntry>[]) List<BloodGlucoseEntry> bloodGlucose,
    @Default(<BloodGlucoseEntry>[]) List<BloodGlucoseEntry> previousBloodGlucose,
    @Default(<BloodGlucoseEntry>[]) List<BloodGlucoseEntry> baselineBloodGlucose,
    @Default(<SkinTemperatureEntry>[])
    List<SkinTemperatureEntry> skinTemperature,
    @Default(<SkinTemperatureEntry>[])
    List<SkinTemperatureEntry> previousSkinTemperature,
    @Default(<SkinTemperatureEntry>[])
    List<SkinTemperatureEntry> baselineSkinTemperature,
    // Long-range (non-day) overview: one aggregated point per day plus the true
    // latest reading, so the year chart and cards never load the raw record
    // list. Empty/null on the day view, which keeps using the raw lists above.
    // See VitalsRepositoryImpl._loadVitalsPeriodRaw (VitalsPeriodMetric.all).
    @Default(<DailyBloodPressurePoint>[])
    List<DailyBloodPressurePoint> bloodPressureDaily,
    @Default(<DailyVitalPoint>[]) List<DailyVitalPoint> spO2Daily,
    @Default(<DailyVitalPoint>[]) List<DailyVitalPoint> respiratoryRateDaily,
    @Default(<DailyVitalPoint>[]) List<DailyVitalPoint> bodyTemperatureDaily,
    @Default(<DailyVitalPoint>[]) List<DailyVitalPoint> vo2MaxDaily,
    @Default(<DailyVitalPoint>[]) List<DailyVitalPoint> bloodGlucoseDaily,
    @Default(<DailyVitalPoint>[]) List<DailyVitalPoint> skinTemperatureDaily,
    BloodPressureEntry? latestBloodPressure,
    SpO2Entry? latestSpO2,
    Vo2MaxEntry? latestVo2Max,
    RespiratoryRateEntry? latestRespiratoryRate,
    BodyTempEntry? latestBodyTemperature,
    BloodGlucoseEntry? latestBloodGlucose,
    SkinTemperatureEntry? latestSkinTemperature,
  }) = _VitalsPeriodData;
}
